src/main/java/de/tum/in/www1/artemis/service/CourseService.java

```diff
@@ -271,6 +274,107 @@ public void registerUserForCourse(User user, Course course) {
+        List<Map<String, Object>> distinctOutcome = removeDuplicatesFromMapList(outcome, startDate);
+        return createResultArray(distinctOutcome, endDate);
+    }
+
+    /**
+     * the List of maps result contains duplicated entries. This method compares the values and returns a List<Map<String, Object>>
+     * without duplicated entries
+     *
+     * @param result the result given by the Repository call
+     * @param startDate the startDate of the period
+     * @return A List<Map<String, Object>> containing date and amount of active users in this period
+     */
+    private List<Map<String, Object>> removeDuplicatesFromMapList(List<Map<String, Object>> result, ZonedDateTime startDate) {
+        Map<Object, List<String>> users = new HashMap<>();
+        for (Map<String, Object> listElement : result) {
+            ZonedDateTime date = (ZonedDateTime) listElement.get("day");
+            int index = getWeekOfDate(date);
+            String username = listElement.get("username").toString();
+            List<String> usersInSameSlot = users.get(index);
+            // if this index is not yet existing in users
+            if (usersInSameSlot == null) {
+                usersInSameSlot = new ArrayList<>();
+                usersInSameSlot.add(username);
+                users.put(index, usersInSameSlot);
+            }   // if the value of the map for this index does not contain this username
+            else if (!usersInSameSlot.contains(username)) {
+                usersInSameSlot.add(username);
+            }
+        }
+        List<Map<String, Object>> returnList = new ArrayList<>();
+        users.forEach((k, v) -> {
+            int year = (Integer) k < getWeekOfDate(startDate) ? startDate.getYear() + 1 : startDate.getYear();
+            ZonedDateTime firstDateOfYear = ZonedDateTime.of(year, 1, 1, 0, 0, 0, 0, startDate.getZone());
+            ZonedDateTime start = getWeekOfDate(firstDateOfYear) == 1 ? firstDateOfYear.plusWeeks(((Integer) k) - 1) : firstDateOfYear.plusWeeks((Integer) k);
+            Map<String, Object> listElement = new HashMap<>();
+            listElement.put("day", start);
+            listElement.put("amount", (long) v.size());
+            returnList.add(listElement);
+        });
+        return returnList;
+    }
+
+    /**
+     * Gets a list of maps, each map describing an entry in the database. The map has the two keys "day" and "amount",
+     * which map to the date and the amount of the findings. This Map-List is taken and converted into a Integer array,
+     * containing the values for each bar of the graph
+     *
+     * @param outcome A List<Map<String, Object>>, containing the content which should be refactored into an array
+     * @param endDate the endDate
+     * @return an array, containing the values for each bar in the graph
+     */
+    private Integer[] createResultArray(List<Map<String, Object>> outcome, ZonedDateTime endDate) {
+        Integer[] result = new Integer[4];
```

Comment:
Why 4? Maybe explain a little bit what graph and what you want at each index.

Reply:
👍